haskell/src/Lentille/GraphQL.hs

```diff
@@ -110,17 +136,55 @@ data RateLimit = RateLimit {used :: Int, remaining :: Int, resetAt :: UTCTime}
 instance From RateLimit Text where
   from RateLimit {..} = "remains:" <> show remaining <> ", reset at: " <> show resetAt
 
+type RetryCheck m a = (Either (FetchError a) a, [ReqLog]) -> m Bool
+
+-- | wrapper around fetchWithLog than can optionaly handle fetch retries
+-- based on the returned data inspection via a provided function (see RetryCheck).
+-- In case of retry the depth parameter of mkArgs is decreased (see adaptDepth)
+doRequest ::
+  forall a.
+  forall m.
+  (MonadGraphQLE m, Fetch a, FromJSON a) =>
+  GraphClient ->
+  (Maybe Int -> Maybe Text -> Args a) ->
+  Maybe (RetryCheck m a) ->
+  Maybe Int ->
+  Maybe PageInfo ->
+  m (Either (FetchError a) a, [ReqLog])
+doRequest client mkArgs retryCheckM depthM pageInfoM = gRetry retryCheck runFetch
+  where
+    gRetry = genericRetry Macroscope "Retrying request with smaller depth"
+    retryCheck _ = fromMaybe (const $ pure False) retryCheckM
+    runFetch :: (MonadGraphQLE m) => Int -> m (Either (FetchError a) a, [ReqLog])
+    runFetch retried = do
+      let adapedDepth = adaptDepth
+      if adapedDepth == Just 0
+        then error "Unable to reduce depth more"
```

Comment:
since error is a bottom, you can also replace the if/then/else with:

```suggestion
        when (adapedDepth == just 0) $ error "Unable to reduce depth more"
```

Reply:
I've set a max to 1 so I'm removing the exception raise here. It's unlikely  to happen that a single PR cannot be fetched.